File: android/src/main/kotlin/id/co/rfbnr/flutter_indocard_ocr/ModuleOCR/model/NPWPModel.kt
package id.co.rfbnr.flutter_indocard_ocr.ModuleOCR.model

import org.json.JSONObject

/**
 * NPWP Data Model
 * Data model for Indonesian Tax ID Card (NPWP)
 */
data class NPWPModel(
    var npwp: String? = null,              // NPWP Number (format: XX.XXX.XXX.X-XXX.XXX)
    var nik: String? = null,               // NIK (KTP Number)
    var nama: String? = null,              // Full Name
    var alamat: String? = null,            // Full Address
    var kpp: String? = null,               // KPP (Kantor Pelayanan Pajak)
    var confidence: Int = 0
) {
    /**
     * Convert to JSON Map
     */
    fun toJson(): Map<String, String> {
        return mapOf(
            "npwp" to (npwp ?: ""),
            "nik" to (nik ?: ""),
            "nama" to (nama ?: ""),
            "alamat" to (alamat ?: ""),
            "kpp" to (kpp ?: "")
        )
    }

    /**
     * Convert to JSON String
     */
    fun toJsonString(): String {
        val dict = toJson()
        val json = JSONObject(dict)
        return json.toString()
    }

    /**
     * Get formatted NPWP (XX.XXX.XXX.X-XXX.XXX)
     */
    fun getFormattedNPWP(): String {
        val cleanNPWP = npwp?.replace(Regex("[^0-9]"), "") ?: return ""

        if (cleanNPWP.length == 15) {
            return "${cleanNPWP.substring(0, 2)}.${cleanNPWP.substring(2, 5)}.${cleanNPWP.substring(5, 8)}.${cleanNPWP.substring(8, 9)}-${cleanNPWP.substring(9, 12)}.${cleanNPWP.substring(12, 15)}"
        }

        return cleanNPWP
    }
}

File: ios/Classes/ModuleOCR/models/NPWPModel.swift
import Foundation

struct NPWPModel {
    var npwp: String?           // NPWP Number (format: XX.XXX.XXX.X-XXX.XXX)
    var nik: String?            // NIK (KTP Number)
    var nama: String?           // Full Name
    var alamat: String?         // Full Address
    var kpp: String?            // KPP (Kantor Pelayanan Pajak)
    var confidence: Int = 0

    init() {
        self.npwp = nil
        self.nik = nil
        self.nama = nil
        self.alamat = nil
        self.kpp = nil
        self.confidence = 0
    }

    // Convert to JSON dictionary
    func toJson() -> [String: String] {
        return [
            "npwp": npwp ?? "00.000.000.0-000.000",
            "nik": nik ?? "unknown",
            "nama": nama ?? "unknown",
            "alamat": alamat ?? "unknown",
            "kpp": kpp ?? "unknown"
        ]
    }

    // Convert to JSON string
    func toJsonString() -> String? {
        let dict = toJson()
        guard let jsonData = try? JSONSerialization.data(withJSONObject: dict, options: []),
              let jsonString = String(data: jsonData, encoding: .utf8) else {
            return nil
        }
        return jsonString
    }

    // Get formatted NPWP (XX.XXX.XXX.X-XXX.XXX)
    func getFormattedNPWP() -> String {
        guard let npwp = npwp else { return "" }
        let cleanNPWP = npwp.replacingOccurrences(of: "[^0-9]", with: "", options: .regularExpression)

        if cleanNPWP.count == 15 {
            let index0 = cleanNPWP.index(cleanNPWP.startIndex, offsetBy: 2)
            let index1 = cleanNPWP.index(cleanNPWP.startIndex, offsetBy: 5)
            let index2 = cleanNPWP.index(cleanNPWP.startIndex, offsetBy: 8)
            let index3 = cleanNPWP.index(cleanNPWP.startIndex, offsetBy: 9)
            let index4 = cleanNPWP.index(cleanNPWP.startIndex, offsetBy: 12)

            let part1 = cleanNPWP[..<index0]
            let part2 = cleanNPWP[index0..<index1]
            let part3 = cleanNPWP[index1..<index2]
            let part4 = cleanNPWP[index2..<index3]
            let part5 = cleanNPWP[index3..<index4]
            let part6 = cleanNPWP[index4...]

            return "\(part1).\(part2).\(part3).\(part4)-\(part5).\(part6)"
        }

        return cleanNPWP
    }
}
